File: fila/FilaDinamica.cpp
#include <iostream>

using namespace std;

struct Celula {
    int elemento;
    Celula *prox;
};

class Fila {
private:
    Celula *prim;
    Celula *ult;
    int tamanho;

public:
    //construtor
    Fila () {
        prim = new Celula;
        prim->elemento=0;
        prim->prox=NULL;
        ult=prim;
        tamanho=0;
    }
    //destrutor
    ~Fila () {
        Celula *aux;
        while (prim != NULL) {
            aux = prim;
            prim = prim->prox;
            delete(aux);
        }
    }

    void enfileirar (int x) {
        ult->prox = new Celula;
        ult = ult->prox;
        ult->elemento = x;
        ult->prox = NULL;
        tamanho++;
    }

    int desenfileirar () {
        Celula *aux = prim->prox;
        int retornar = aux->elemento;
        prim->prox = prim->prox->prox;
        delete(aux);
        if (prim->prox == NULL) {
            ult = prim;
        }
        tamanho--;
        return retornar;
    }

    void mostrarFila () {
        Celula *aux;
        while (prim != NULL) {
            aux = prim;
            printf("%d", aux->elemento);
            aux = aux->prox;
        }
        aux = NULL;
    }
    /*
    int mostrarTamanho
    bool vazia
*/

};

File: fila/FilaDinamica_java/FilaDinamica.java
package FilaDinamica_java;

public class FilaDinamica {    
    private Celula prim;
    private Celula ult;
    private int tamanho;
    
    //construtor
    public FilaDinamica () {
        prim = new Celula();
        prim.elemento=0;
        ult=prim;
        tamanho=0;
    }

    //destrutor
    // public ~FilaDinamica () {
    //     Celula aux;
    //     while (prim != null) {
    //         aux = prim;
    //         prim = prim.prox;
    //         aux = null;
    //     }
    // }

    public void inserir (int x) {
        ult.prox = new Celula();
        ult = ult.prox;
        ult.elemento = x;
        ult.prox = null;
        tamanho++;
    }

    public int remover () throws Exception{
        if (prim.prox == null) {
            throw new Exception("Erro");
        }
        Celula aux = prim;
        int retornar = aux.elemento;
        prim = prim.prox;
        aux.prox = null;
        aux = null;
        tamanho--;
        return retornar;
    }

    public void mostrarFila () {
        Celula aux;
        aux = prim;
        while (aux != null) {
            System.out.printf("%d", aux.elemento);
            aux = aux.prox;
        }
        aux = null;
    }

    int getTamanho () {
        return tamanho;
    }
    /*
    bool vazia
    s*/
}
